storage/postgres/query_builder.go

```diff
@@ -44,336 +84,328 @@ func NewQueryBuilder(db pgDB) *QueryBuilder {
-		returningSQL().
-		mergeQueryParams()
-
-	if pgq.err != nil {
-		return pgq.err
+	for _, criterion := range criteria {
+		if err := criterion.Validate(); err != nil {
+			pq.err = err
+			return pq
+		}
+		if hasMultiVariateOp(criteria) {
+			pq.shouldRebind = true
+		}
+		switch criterion.Type {
+		case query.FieldQuery:
+			columns := columnsByTags(pq.entityTags)
+			if !columns[criterion.LeftOp] {
+				pq.err = &util.UnsupportedQueryError{Message: fmt.Sprintf("unsupported field query key: %s", criterion.LeftOp)}
+				return pq
+			}
+			pq.fieldsWhereClause.children = append(pq.fieldsWhereClause.children, &whereClauseTree{
+				criterion: criterion,
+				dbTags:    pq.entityTags,
+				tableName: pq.entityTableName,
+			})
+		case query.LabelQuery:
+			pq.labelsWhereClause.children = append(pq.labelsWhereClause.children, &whereClauseTree{
+				operator: AND,
+				children: []*whereClauseTree{
+					{
+						criterion: query.ByField(query.EqualsOperator, "key", criterion.LeftOp),
+						dbTags:    pq.labelEntityTags,
+					},
+					{
+						criterion: query.ByField(criterion.Operator, "val", criterion.RightOp...),
+						dbTags:    pq.labelEntityTags,
+					},
+				},
+			})
+		case query.ResultQuery:
+			pq.processResultCriteria(criterion)
+		}
 	}
 
-	sql := pgq.sql.String()
-	pgq.sql.Reset()
-	pgq.sql.WriteString(pgq.db.Rebind(sql))
-	pgq.sql.WriteString(";")
+	return pq
+}
 
-	log.C(ctx).Debugf("Executing postgres query: %s", pgq.sql.String())
-	return nil
+func (pq *pgQuery) WithLock() *pgQuery {
+	if pq.err != nil {
+		return pq
+	}
+	if _, ok := pq.db.(*sqlx.Tx); ok {
+		pq.hasLock = true
+	}
+	return pq
 }
 
-func (pgq *pgQuery) orderBySQL() *pgQuery {
-	if sql, err := orderBySQL(pgq.orderByFields); err != nil {
-		pgq.err = err
-		return pgq
-	} else {
-		pgq.sql.WriteString(sql)
+func (pq *pgQuery) limitSQL() string {
+	if len(pq.limit) > 0 {
+		pq.queryParams = append(pq.queryParams, pq.limit)
+		return "LIMIT ?"
 	}
-	return pgq
+	return ""
 }
 
-func (pgq *pgQuery) returningSQL() *pgQuery {
-	for i := range pgq.returningFields {
-		if !strings.HasPrefix(pgq.returningFields[i], fmt.Sprintf("%s.", mainTableAlias)) {
-			pgq.returningFields[i] = fmt.Sprintf("%s.%s", mainTableAlias, pgq.returningFields[i])
-		}
+func (pq *pgQuery) whereSQL() string {
+	whereClause := &whereClauseTree{
+		operator: AND,
+		children: []*whereClauseTree{
+			pq.fieldsWhereClause,
+			pq.labelsWhereClause,
+		},
+	}
+	whereSQL, queryParams := whereClause.compileSQL()
+	if len(whereSQL) == 0 {
+		return ""
 	}
+	pq.queryParams = append(pq.queryParams, queryParams...)
+	return fmt.Sprintf(" WHERE %s", whereSQL)
+}
 
-	if len(pgq.returningFields) == 1 {
-		pgq.sql.WriteString(fmt.Sprintf(" RETURNING " + pgq.returningFields[0]))
-	} else if len(pgq.returningFields) > 0 {
-		pgq.sql.WriteString(" RETURNING " + strings.Join(pgq.returningFields, ", "))
+func (pq *pgQuery) returningSQL() string {
+	fieldsCount := len(pq.returningFields)
+	switch fieldsCount {
```

Comment:
The whole switch is redundant since this is handled in `string.Join(...)`
```
func Join(a []string, sep string) string {
	switch len(a) {
	case 0:
		return ""
	case 1:
		return a[0]
...
```